apps/src/music/views/ExemplarPlayerView.tsx

```diff
@@ -0,0 +1,126 @@
+import FontAwesomeV6Icon from '@code-dot-org/component-library/fontAwesomeV6Icon';
+import classNames from 'classnames';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+
+import {useAppDispatch, useAppSelector} from '@cdo/apps/util/reduxHooks';
+
+import {DEFAULT_PACK} from '../constants';
+import {PlaybackEvent} from '../player/interfaces/PlaybackEvent';
+import MusicLibrary from '../player/MusicLibrary';
+import MusicPlayer from '../player/MusicPlayer';
+import Simple2Sequencer from '../player/sequencer/Simple2Sequencer';
+import {setIsPlaying} from '../redux/musicRedux';
+
+import moduleStyles from './ExemplarPlayer.module.scss';
+
+interface ExemplarPlayerViewProps {
+  getPlaybackEvents: () => PlaybackEvent[];
+  title: string;
+  packId: string | null;
+}
+
+const ExemplarPlayerView: React.FunctionComponent<ExemplarPlayerViewProps> = ({
+  getPlaybackEvents,
+  title,
+  packId,
+}) => {
+  const dispatch = useAppDispatch();
+  const isPlaying = useAppSelector(state => state.music.isPlaying);
+
+  const playerRef = useRef<MusicPlayer | null>(null);
+  if (playerRef.current === null) {
+    playerRef.current = new MusicPlayer();
+  }
+  const simple2SequencerRef = useRef<Simple2Sequencer>(new Simple2Sequencer());
+  const [exemplarIsPlaying, setExemplarIsPlaying] = useState<boolean>(false);
+  // Adjust the packId and preload sounds.
+  const onMount = useCallback(async () => {
+    const currentLibrary = MusicLibrary.getInstance();
+    if (currentLibrary) {
+      currentLibrary.setCurrentPackId(packId);
+      playerRef.current?.updateConfiguration(
+        currentLibrary.getBPM(),
+        currentLibrary.getKey()
+      );
+    }
+
+    // Clear any prior events to prepare playback.
+    simple2SequencerRef.current.clear();
+    await playerRef.current?.preloadSounds(getPlaybackEvents(), () => {});
+  }, [packId, getPlaybackEvents]);
+
+  useEffect(() => {
+    onMount();
+  }, [onMount]);
+
+  // Uses the already compiled song to preload sounds and play it.
+  const onPlaySong = useCallback(async () => {
+    Blockly.getMainWorkspace().hideChaff();
+    dispatch(setIsPlaying(false));
+    playerRef.current?.stopSong();
+
+    // Play the song using the compiled events.
+    playerRef.current?.playSong(getPlaybackEvents());
+
+    setExemplarIsPlaying(true);
+  }, [dispatch, getPlaybackEvents]);
+
+  // Stop the exemplar song, updating Redux and local state.
+  const onStopSong = useCallback(() => {
+    if (!isPlaying) {
+      // If the player was stopped by the Run button, we do not want to interfere with
+      // MusicLabView's control of the player.
+      playerRef.current?.stopSong();
+    }
```

Comment:
Want to make sure I understand this; if we click stop on the exemplar player, and the main lab music player was _not_ playing, then we don't stop the song here? Wouldn't we still want to?

Reply:
This is actually for when the Run button is used to interrupt the exemplar being played. There’s some async stuff happening so when we reach this point, the main timeline has already begun playing the student project. Since it’s the same player, stopping here would actually result in stopping playback of the student song.

Reply:
Ahh got it; assuming this is because even though we have different `player` instances, we're manipulating the same core ToneJS transport system